smolyay/surrogate.py

```diff
@@ -1,546 +1,717 @@
+        if not y0 is None:
+            # validate data inputs
+            if X0 is None:
+                X0 = self.domain[:, 0].reshape((1, -1))
+            else:
+                X0 = numpy.array(X0, ndmin=2)
+            if X0.shape != (1, self.num_dimensions):
+                raise IndexError(
+                    "Must be 2D array with shape (1, num_dimensions) if not None."
+                )
+            y0 = numpy.array(y0).item(0)
+            predicted_y = self.predict(X0)
+            integration_constant = y0 - predicted_y
+            self._integration_constant = integration_constant
+        return self
+
+    @abc.abstractmethod
+    def _create_terms(self):
+        """create terms for the surrogate"""
+        pass
+
+
+class TensorProductSurrogate(SetProductSurrogate):
+    """A surrogate that uses a combination of tensor product as terms
 
-    def _reset_grid(self):
-        """Reset the grids, coefficients and points."""
-        self._grid = None
-        self._coefficients = None
-        self._points = None
-        self._data = None
+    Parameters
+    ----------
+    domain: numpy.ndarray
+        the domain of the surrgoate
 
+    basis_set: BasisFunctionSet or list of BasisFunctionSet
+        the set of basis functions used to combine terms
 
-class GradientSurrogate(Surrogate):
-    r"""Create a surrogate to approximate a complex function.
+    regularization : [:class:L1Regularization, :class:L2Regularization, None], default None
+        the regularization method for determining the coefficients
+    """
 
-    `domain`, `grid_generator` can be defined as in :class:`Surrogate`.
-    The basis matrix ((number of grid points*dimension)
-    x number of grid points) is generated by evaluating the gradients of
-    basis functions at grid points in all dimension (A).
-    The gradient of the function is then computed at all points (B).
-    By solving the overdetermined systems of equations
-    through least-square regression, the coefficients of the surrogate
-    are evaluated in the :meth:`train`, using the gradient of a function
-    as the input. Here, the callable should return the gradient, meaning
-    that output size should be equal to the dimensionality for a given point
-    in the domain. Alternatively, surrogate can be trained provided the
-    gradient function's value at grid points
-    (size = number of grid points x dimension) through :meth:`train_from_data`.
-    One can evaluate the surrogate at a given point through :meth:`__call__`.
-    :meth:`gradient` takes the gradeint of the surrogate at a given point.
+    def _create_terms(self):
+        num_points = numpy.prod([len(p) for p in self._basis_sets])
+
+        self._index_combinations = numpy.zeros(
+            (num_points, len(self._basis_sets)), dtype=int
+        )
+        for i, point in enumerate(
+            itertools.product(*[numpy.arange(len(bs)) for bs in self._basis_sets])
```

Comment:
Will this work with `range`? `arange` fully allocates the array, but `range` should just be an iterable so this will reduce memory consumption.
```suggestion
            itertools.product(*[range(len(bs)) for bs in self._basis_sets])
```